client-spark/spark-3/src/main/scala/org/apache/spark/shuffle/celeborn/CelebornShuffleReader.scala

```diff
@@ -62,15 +70,51 @@ class CelebornShuffleReader[K, C](
         metrics.incFetchWaitTime(time)
     }
 
+    if (streamCreatorPool == null) {
+      CelebornShuffleReader.synchronized {
+        if (streamCreatorPool == null) {
+          streamCreatorPool = ThreadUtils.newDaemonCachedThreadPool(
+            "create-stream-thread",
+            conf.readStreamCreatorPoolThreads,
+            60);
+        }
+      }
+    }
+
+    val streams = new ConcurrentHashMap[Integer, CelebornInputStream]()
+    (startPartition until endPartition).map(partitionId => {
+      streamCreatorPool.submit(new Runnable {
+        override def run(): Unit = {
+          if (exceptionRef.get() == null) {
+            try {
+              val inputStream = shuffleClient.readPartition(
+                handle.shuffleId,
+                partitionId,
+                context.attemptNumber(),
+                startMapIndex,
+                endMapIndex)
+              streams.put(partitionId, inputStream)
+            } catch {
+              case e: IOException =>
+                logInfo("Exception caught when readPartition!")
+                exceptionRef.compareAndSet(null, e)
+            }
+          }
+        }
+      })
+    })
+
     val recordIter = (startPartition until endPartition).iterator.map(partitionId => {
       if (handle.numMappers > 0) {
         val start = System.currentTimeMillis()
-        val inputStream = shuffleClient.readPartition(
-          handle.shuffleId,
-          partitionId,
-          context.attemptNumber(),
-          startMapIndex,
-          endMapIndex)
+        var inputStream: CelebornInputStream = streams.get(partitionId)
+        while (inputStream == null) {
+          if (exceptionRef.get() != null) {
```

Comment:
there is a chance that shuffleClient.readPartition throws a non-IOException, then it goes dead lock

Reply:
+1

Reply:
the method `shuffleClient.readPartition` explicitly declares the throwing of an IOException.

https://github.com/apache/incubator-celeborn/blob/b66eaff880e91864a21c96dbac94fa5f8cd84f4c/client/src/main/java/org/apache/celeborn/client/ShuffleClient.java#L191-L193

Reply:
> the method `shuffleClient.readPartition` explicitly declares the throwing of an IOException.
> 
> https://github.com/apache/incubator-celeborn/blob/b66eaff880e91864a21c96dbac94fa5f8cd84f4c/client/src/main/java/org/apache/celeborn/client/ShuffleClient.java#L191-L193

Right. Do we have to consider RuntimeException here?

Reply:
should catch Throwable for safe